controllers/baseboardmanagement_controller.go

```diff
@@ -18,42 +18,233 @@ package controllers
+	// fieldReconcilers defines BaseboardManagement spec field reconciler functions
+	fieldReconcilers := []baseboardManagementFieldReconciler{
+		r.reconcilePower,
+	}
+	for _, reconiler := range fieldReconcilers {
+		if err := reconiler(ctx, bm, bmcClient); err != nil {
+			logger.Error(err, "Failed to reconcile BaseboardManagement", "host", bm.Spec.Connection.Host)
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Patch the status after each reconciliation
+	return r.reconcileStatus(ctx, bm)
+}
+
+func (r *BaseboardManagementReconciler) reconcilePower(ctx context.Context, bm *bmcv1alpha1.BaseboardManagement, bmcClient BMCClient) error {
+	powerStatus, err := bmcClient.GetPowerState(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get power state: %v", err)
+	}
+
+	// If BaseboardManagement has desired power state then return
+	if bm.Spec.Power == bmcv1alpha1.PowerState(strings.ToLower(powerStatus)) {
+		return nil
+	}
 
-	// TODO(user): your logic here
+	// Setting baseboard management to desired power state
+	_, err = bmcClient.SetPowerState(ctx, string(bm.Spec.Power))
+	if err != nil {
+		return fmt.Errorf("failed to set power state: %v", err)
+	}
+
+	return nil
+}
+
+// setCondition updates the status.Condition if the condition type is present.
+// Appends if new condition is found.
+// Patches the BaseboardManagement status.
+func (r *BaseboardManagementReconciler) setCondition(ctx context.Context, bm *bmcv1alpha1.BaseboardManagement, cType bmcv1alpha1.BaseboardManagementConditionType, message string) (ctrl.Result, error) {
+	patch := client.MergeFrom(bm.DeepCopy())
+
+	currentConditions := bm.Status.Conditions
+	for i := range currentConditions {
+		// If condition exists, update the message if different
+		if currentConditions[i].Type == cType {
+			if currentConditions[i].Message != message {
+				bm.Status.Conditions[i].Message = message
+				return r.patchStatus(ctx, bm, patch)
+			}
+			return ctrl.Result{}, nil
+		}
+	}
+
+	// Append new condition to Conditions
+	condition := bmcv1alpha1.BaseboardManagementCondition{
+		Type:    cType,
+		Message: message,
+	}
+	bm.Status.Conditions = append(bm.Status.Conditions, condition)
+
+	return r.patchStatus(ctx, bm, patch)
+}
+
+// reconcileStatus updates the Power and Conditions and patches BaseboardManagement status.
+func (r *BaseboardManagementReconciler) reconcileStatus(ctx context.Context, bm *bmcv1alpha1.BaseboardManagement) (ctrl.Result, error) {
+	patch := client.MergeFrom(bm.DeepCopy())
+
+	// Update the power status of the BaseboardManagement
+	bm.Status.Power = bm.Spec.Power
+	// Clear conditions
+	bm.Status.Conditions = []bmcv1alpha1.BaseboardManagementCondition{}
+
+	return r.patchStatus(ctx, bm, patch)
+}
+
+// patchStatus patches the specifies patch on the BaseboardManagement.
+func (r *BaseboardManagementReconciler) patchStatus(ctx context.Context, bm *bmcv1alpha1.BaseboardManagement, patch client.Patch) (ctrl.Result, error) {
+	err := r.client.Status().Patch(ctx, bm, patch)
```

Comment:
@chrisdoherty4 @jacobweinstock there is no idempotency/locking in the whole reconcile, do we want/need that?

Reply:
Nope. When you're only considering the object under reconciliation there's no races as the manager ensures you're the only reconciler reconciling that resource.

Reply:
There is, of course, the possibility that an object is patched by something else but I don't think we have those semantics here.